src/eos-updater-branch-file.c

```diff
@@ -0,0 +1,309 @@
+                                  contents_len,
+                                  G_KEY_FILE_NONE,
+                                  error))
+    return NULL;
+
+  sha512 = g_checksum_new (G_CHECKSUM_SHA512);
+  g_checksum_update (sha512, contents_bytes, contents_len);
+  bf->contents_sha512sum = g_strdup (g_checksum_get_string (sha512));
+
+  if (signature != NULL)
+    {
+      if (!eos_updater_get_timestamp_from_branch_file_keyfile (bf->branch_file,
+                                                               &bf->download_time,
+                                                               error))
+        return NULL;
+    }
+  else
+    bf->download_time = g_date_time_ref (download_time);
+
+  return g_steal_pointer (&bf);
+}
+
+gboolean
+eos_branch_file_save_to_repo (EosBranchFile *branch_file,
+                              OstreeRepo *repo,
+                              GCancellable *cancellable,
+                              GError **error)
+{
+  g_autoptr(GFile) ext_dir = get_eos_extensions_dir (repo);
+  g_autoptr(GFile) branch_file_path = NULL;
+  g_autoptr(GFile) signature_path = NULL;
+
+  get_branch_file_paths (ext_dir, &branch_file_path, &signature_path);
+  return eos_branch_file_save (branch_file,
+                               branch_file_path,
+                               signature_path,
+                               cancellable,
+                               error);
+}
+
+gboolean
+eos_branch_file_save (EosBranchFile *branch_file,
+                      GFile *target,
+                      GFile *target_signature,
+                      GCancellable *cancellable,
+                      GError **error)
+{
+  gconstpointer contents;
+  gsize len;
+
+  g_return_val_if_fail (EOS_IS_BRANCH_FILE (branch_file), FALSE);
+
+  contents = g_bytes_get_data (branch_file->raw_contents,
+                               &len);
+  if (!g_file_replace_contents (target,
+                                contents,
+                                len,
+                                NULL,
+                                FALSE,
+                                G_FILE_CREATE_NONE,
+                                NULL,
+                                cancellable,
+                                error))
+    return FALSE;
+
+  if (branch_file->raw_signature == NULL)
+    {
+      gint64 unix_time = g_date_time_to_unix (branch_file->download_time);
+
+      if (unix_time < 0)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                       "invalid download time of the branch file");
+          return FALSE;
+        }
+      if (!g_file_set_attribute_uint64 (target,
+                                        G_FILE_ATTRIBUTE_TIME_MODIFIED,
+                                        unix_time,
+                                        G_FILE_QUERY_INFO_NONE,
+                                        cancellable,
+                                        error))
```

Comment:
Is there any reason why we're changing the mtime of the target to the branch file's download time?